include/chemfiles/files/GzFile.hpp

```diff
@@ -0,0 +1,53 @@
+// Chemfiles, a modern library for chemistry file reading and writing
+// Copyright (C) Guillaume Fraux and contributors -- BSD license
+
+#ifndef CHEMFILES_GZ_FILES_HPP
+#define CHEMFILES_GZ_FILES_HPP
+
+#include <streambuf>
+#include <vector>
+
+#include <zlib.h>
+
+#include "chemfiles/File.hpp"
+
+namespace chemfiles {
+
+class gzstreambuf: public std::streambuf {
```

Comment:
Is this used exclusively by `GzFile`? Can `GzFile` only support TextFiles? I'm asking as binary formats can be compressed as well so I want to make sure the naming conventions are appropriate here.

Most compressed binary formats are read by reading the uncompressed data into the buffer, as is done here, and then passing that buffered data to the appropriate library for reading. I am trying to suggest a way to make this streamlined for our purposes, or the binary formats can handle compressed data however they wish and a unified interface only exists for text data.

Maybe it would be good to make a single buffer class that holds both text and binary data and other classes can read this buffer directly or line by line in the case of a text file. This is probably above the needs of this PR though - just brain storming.

Reply:
> Is this used exclusively by GzFile? Can GzFile only support TextFiles? 

Yes and yes, for now.

This is an implementation of a std::streambuf for reading gz files. The whole interface is built for text-like files, and only decompress parts of the file as needed. This works using the same interface as usual C++ stream objects.

What you mean is about compressed binary formats, like .xtc.gz ? In this case we would need to decompress the whole file before reading it from memory. I don't think there is a generic solution for this. Most binary formats I know already use some kind of internal compression, and in this case there is not much gain to get by compressing the file again.

Also, we would need a binary library that can read from a buffer in memory, which is not the case for all of them.

I think this could be interesting, but it needs to be decided on a format by format basis, as there is no common way to do it. We could still use `gzstreambuf/xzstreambuf` to implement reading and decompressing the whole file in memory (basically calling underflow again and again, copying the resulting values in the buffer).

Reply:
That comment was more idea "vomit" than anything. We would definitely need to read the entire file into memory for most decoders unless they provide their own method of decompressing the files. I agree that it is unlikely that we cannot find a "one size fits all" solution for this issue. Is there a way for the non-text file classes to know that their are getting a compressed file? For example, XTC files will probably be handled by the XDR file class (when ported from molfiles, XDR is common to XTC and TRR). How would XDR now to decompress and feed the binary data to libxdr?

Reply:
> Is there a way for the non-text file classes to know that their are getting a compressed file?

No, not really. They could check the file name for a different extension though.

Concerning XRD, I don't know how the format works. Is the whole file compressed? This does not sound efficient from a MD point of view: one need to be able to append more data every so often. Or are you talking about compressions realized by the user after the file was created by another software ? In this case, I think the best thing to do is to dispatch on file extension, like `TextFile::create` does today.